pkg/migration/migration_util.go

```diff
@@ -0,0 +1,163 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package migration
+
+import (
+	"fmt"
+	"strings"
+
+	"persistent-volume-migrator/internal/ceph/rbd"
+	"persistent-volume-migrator/pkg/k8sutil"
+
+	logger "persistent-volume-migrator/pkg/log"
+
+	v1 "k8s.io/api/core/v1"
+	k8s "k8s.io/client-go/kubernetes"
+)
+
+// migratePVC migrates a list of PVCs to CSI.
+func migratePVC(client *k8s.Clientset, pvc v1.PersistentVolumeClaim, destinationStorageClass,
+	rookNamespace, cephClusterNamespace string) error {
+
+	logger.DefaultLog("migrating PVC %v", pvc.Name)
+
+	logger.DefaultLog("Fetch PV information from PVC %s", pvc.Name)
+	pv, err := k8sutil.GetPV(client, pvc.Spec.VolumeName)
+	if err != nil {
+		return fmt.Errorf("failed to get PV object with name %s: %v", pvc.Spec.VolumeName, err)
+	}
+	logger.DefaultLog("PV found %v ", pv)
+
+	logger.DefaultLog("Update Reclaim policy from Delete to Reclaim for PV: %s", pv.Name)
+	err = k8sutil.UpdateReclaimPolicy(client, pv)
+	if err != nil {
+		return fmt.Errorf("failed to update ReclaimPolicy for PV object %s: %v", pvc.Spec.VolumeName, err)
+	}
+
+	logger.DefaultLog("Retrieving old ceph volume name from PV object: %s", pv.Name)
+	rbdImageName := k8sutil.GetFlexVolumeName(pv)
+	if rbdImageName == "" {
+		return fmt.Errorf("rbdImageName cannot be empty in Flex PV object: %v", pv)
+	}
+	logger.DefaultLog("Flex rbdImageName name %v ", rbdImageName)
+
+	logger.DefaultLog("Deleting pvc object: %s", pvc.Name)
+	err = k8sutil.DeletePVC(client, &pvc) // nolint:gosec // skip gosec as pvc is accessed via it's reference.
+	if err != nil {
+		// TODO have an option to revert back if some Delete fails?
+		return fmt.Errorf("failed to Delete PVC object %s: %v", pvc.Name, err)
+	}
+
+	logger.DefaultLog("Generate new PVC with same name in destination storageclass")
+	csiPVC := k8sutil.GenerateCSIPVC(destinationStorageClass, &pvc) // nolint:gosec // skip gosec as pvc is accessed via it's reference.
+	logger.DefaultLog("Structure of the generated PVC in destination Storageclass %v ", csiPVC)
+
+	logger.DefaultLog("Create new csi pvc")
+	csiPV, err := k8sutil.CreatePVC(client, csiPVC, 5)
+	if err != nil {
+		return fmt.Errorf("failed to Create CSI PVC object %s: %v", pvc.Name, err)
+	}
+	logger.DefaultLog("New PVC with same name created in CSI. Persistent Volume: %v ", csiPV)
+
+	logger.DefaultLog("Extracting new volume name from CSI PV")
+	csiRBDImageName := k8sutil.GetCSIVolumeName(csiPV)
```

Comment:
Does this helper retry? There will be some delay between the PVC request and the PV being created.

Reply:
Added a logic to retry in case PV is not yet created.